src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs

```diff
@@ -48,14 +65,25 @@ public struct CodeRefactoringContext
         /// Add supplied <paramref name="action"/> to the list of refactorings that will be offered to the user.
         /// </summary>
         /// <param name="action">The <see cref="CodeAction"/> that will be invoked to apply the refactoring.</param>
-        public void RegisterRefactoring(CodeAction action)
+        public void RegisterRefactoring(CodeAction action) => RegisterRefactoring(action, applicableToSpan: default);
```

Comment:
this seems... 'wrong' to me.  If there is no span provided by the actual refactoring, would we not just default to this.Span here?  because, effectively, the refactoring is def saying "i apply to the span you passed in" :)   

Isn't the optional parameter only if the refactoring wants to override that with something more specific from context?

Reply:
Also, if we can manage to do this, then we'll be in the state where there's no need for nullable-spans at all.  Because we'll always have a valid span to consider.

Reply:
Note: 
- `RegisterRefactoring(CodeAction action)` shouldn't pass default but `null` (oversight after switching to `TextSpan?`)

I somehow absolutely missed that there's `this.Span` on that type. I'll need to check what it gets set to (I actually have no idea).

Reply:
"this.Span" is literally the span the user has selected (or hte position the caret is at if there's no selection).  It's what all refactorings *start* with to determine if they should show :)  i.e. in all ComputeRefactoringsAsync things generally start with:

```c#
        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
        {
            var (document, span, cancellationToken) = context;
```

`this.Span` is the span the refactoring starts examining.  So, in general, if the refactoring itself doesn't report a better 'applicabilitySpan' it seems reasonable to start with the one we know it was examining.

Reply:
Ah, right, somehow I didn't realize this is the same Context class (long day :)). 

As for using that, that would cause these refactorings (the ones who didn't opt-in) always closest while I somehow feel we might want to do the very opposite / keep them in their original ordering by the attribute (which is what happens if they specify null right now).

Reply:
you could have a bool if it was the 'default' span that came along, or it was explicitly provided, and order accordingly :)   i'm honestly not sure what the best thing is here.  We'll likely have to try out a few things to see what feels right :)

Reply:
TBH it doesn't really matter much. 

We actually can't pass null properly without leaking the `TextSpan?` as part of public API. The registration lambda gets passed as part in (public) constructor and it needs to be accept whatever we want to put to our (internal) `ImmutableArray<…> CodeActions` storage.

But that means we just can't call the lambda with `null` in the original `RegisterRefactoring()` overload.

And having constructor that takes register lambda accepting `TextSpan?` but having `RegisterRefactoring` method that can only take `TextSpan` is just weird.

Thus, I'd just go with passing default to make them "furthest" instead of `this.Span` and thus making them closest.

Reply:
> We actually can't pass null properly without leaking the TextSpan? as part of public API. The registration lambda gets passed as part in (public) constructor and it needs to be accept whatever we want to put to our (internal) ImmutableArray<…> CodeActions storage.

Can you point to what you mean?

Reply:
> Thus, I'd just go with passing default

This doesn't work in the case of refactorings at the start of a fiel :) (which do exist) :)

Reply:
> This doesn't work in the case of refactorings at the start of a fiel :) (which do exist) :)

I don't follow.

> Can you point to what you mean?

Imagine [this](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-a2c30b83123c3cb46ac84dae8175bc4bR24) accepts null and we want to pass `null` [here](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-4b18194545a98c011598321cbfa1dd9fR68) (carries info "didn't register specific TextSpan") .  The `RegisterRefactoring` [just calls](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-4b18194545a98c011598321cbfa1dd9fL86) `_registerRefactoring` lambda with given parameters and so if we want to call it with `null` the [lambda](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-4b18194545a98c011598321cbfa1dd9fR30) needs to be of type `Action<CodeAction, TextSpan?>` (has to accept nullable `TextSpan`).

The lambda, however, isn't an implementation detail. It gets set in [*public* constructor](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-4b18194545a98c011598321cbfa1dd9fR52). And that leaks the information that the registration should somehow work with Nullable `applicableToSpan` 🤷‍♂️.

It's not terrible, nobody outside probably uses the constructor. But it _just feels weird_. Either have it nullable explicitly in `RegisterRefactoring` (but that communicates wrong idea about it being optional) or have it nowhere.

> Note: we [still support](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-4b18194545a98c011598321cbfa1dd9fR45) the original constructor without `TextSpan applicableToSpan` parameter

Reply:
Generally, going with `default` is wrong because it doesn't even have to be the furthest (it's a valid `TextSpan(0, 0)`, right? But it's not wrong enough for it to matter. The worst case scenario is we'll order something slightly lower/higher than we should which IMO isn't all that consequential. Especially given we haven't even order things at all right now.

Reply:
> I don't follow.

Passing default is just passing the `[0,0)` span.  But that span has meaning, and it would be indistinguishable from someone actually truly passing that span explicitly.

Reply:
> The lambda, however, isn't an implementation detail. It gets set in public constructor. 

?

From your own link: `internal CodeRefactoringContext(`

It's not a public constructor.  It's internal.  We can do whatever we want with it, up to and including making it take a `TextSpan?` or making it take a `TextSpan+bool`.

Basically, it seems to me that we have 100% control of the end to end here.  So we can just pass along the information as accurately as possible.  This can then be used by the end suggestion source in whatever way it sees fit.

So, personally, i would the information to pass should be:

1. the code action.
2. the applicable span for the code action (either explicitly provided by user code, or implicitly provided by the context).
3. a bool to say if this was explicitly provided or not.

This change can be made entirely in the internal surface area of this library.

The final suggestion source can then use this as it sees fit to order appropriately.  There is no ambiguity.  No overlapping implicit and explicit values.  etc. etc. 

Thoughts?

Reply:
> From your own link: internal CodeRefactoringContext(

You're absolutely right. I don't know why I was _convinced_ it's public struct. I'm for nullability instead of <bool, TextSpan>.

> Passing default is just passing the [0,0) span. But that span has meaning, and it would be indistinguishable from someone actually truly passing that span explicitly.

Ah, yep as said in the comment just above I didn't consider that being such an issue. 

Will update it once more (let's hope for the last time this time :D )

Reply:
Wait, it's internal for now but it will (hopefully) be exposed to be public soon. It's a public struct.

Reply:
THB the more I think about it, the more I'm leaning towards going the `TextSpan?` route and the fact that the lambda in (eventually *public*) constructor will accept `TextSpan` isn't all that bad leakage of implementation detail.

Reply:
Thoughts @CyrusNajmabadi @heejaechang ?

Reply:
Don't do this.  The intent is that hte structs are temporary context objects created by the engine.  It's not intended that they be publicly constructable.  this is a super common pattern we use all over the place *including* for public APIs.

Reply:
> Just as the RegisterCodeAction overload for TextSpan is currently internal only for now but will (hopefully) be made public (once we all agree it's stable, that's why I want to do it in two stages).

Making this new method public in the future is fine.  It's expected that 3rd parties would use it.  That's different from the context-struct object.  These are *often* *intentionally* internal as they serve to bridge between the context object and the owning internal coordinator object.  Users will never create these themselves, and things like the lambda/delegate are 100% internal implementation details we would never expose to them.

Reply:
But it already is publicly creatable: http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeRefactorings/CodeRefactoringContext.cs,35

And having a public constructor that is obsolete and then internal one that people should use instead (but can't) doesn't make sense.

Reply:
Sure, the struct should've been an implementation detail not construtable by user but it is and we have to deal with it somehow conistently.

Or am I missing something (I feel like I might be)?

Reply:
> like the lambda/delegate are 100% internal implementation details we would never expose to them. 

That's the issue, it has already happened. 

> Just a nitpic: GitHub has serious delay between you writing the message & me seeing it, hence why my responses are sort of weird (as if I haven't read what you wrote)

Reply:
*son of a*!

Reply:
You see the problem now and while all of my proposed solutions are sub-optimal?

Reply:
Resolution from gitter: 
 > there's a strong arg to be made for just obsoleting those guys and just keeping the new guys internal

> we have precedence for this.